### Assignmnet 2 - Les miserables dataset
### Group: Abdelmalek Hajjam, Monu Chacko

Les miserables dataset is taken from: http://networkdata.ics.uci.edu/data/lesmis/lesmis.gml
Les Misérables is a French historical novel by Victor Hugo, first published in 1862, that is considered one of the greatest novels of the 19th century.
A link is made between two nodes (characters) if they co-appear in the novel. The node size reflects the number of connections each character has to other characters.
The ex-convict, Jean Valjean, is the novel’s central character, hence makes for the largest node. As he spends a great deal of time in the novel running away from Inspector Javert, he is a cluster unto himself. He is also closely tied to his adopted daughter, Cosette, and her future husband, Marius.
#### If one were interested in influencing Jean Valjean, once could target Cosette or Marius if Valjean cannot be directly targeted.

Here is the link to the video: https://youtu.be/qKpMFhubQ5g


In [7]:
import networkx as nx
import matplotlib.pyplot as plt
import nxviz as nv
# import requests
# import io
from pyvis import network as net
%matplotlib inline

In [8]:
#load the gml file
file = 'lesmis.gml'
G = nx.read_gml(file)

# network dataset info
print(nx.info(G))
print('Graph Diameter:', nx.diameter(G))

Name: 
Type: Graph
Number of nodes: 77
Number of edges: 254
Average degree:   6.5974
Graph Diameter: 5


## How do we evaluate the importance of some individuals in a network?

The number of other nodes that one node is connected to is a measure of its centrality. NetworkX implements a degree centrality, which is defined as the number of neighbors that a node has normalized to the number of individuals it could be connected to in the entire graph. This is accessed by using nx.degree_centrality(G).
Degree Centrality shows how many edges/connections a node has.

In [10]:
numNodes = len(G.nodes())
print("# Nodes: ",numNodes)
print("# Edges: ", len(G.edges()))

#return the list of nodes and their degree of centrality <nx.degree_centrality(G)>
list(nx.degree_centrality(G).items())

# Nodes:  77
# Edges:  254


[('Myriel', 0.13157894736842105),
 ('Napoleon', 0.013157894736842105),
 ('MlleBaptistine', 0.039473684210526314),
 ('MmeMagloire', 0.039473684210526314),
 ('CountessDeLo', 0.013157894736842105),
 ('Geborand', 0.013157894736842105),
 ('Champtercier', 0.013157894736842105),
 ('Cravatte', 0.013157894736842105),
 ('Count', 0.013157894736842105),
 ('OldMan', 0.013157894736842105),
 ('Labarre', 0.013157894736842105),
 ('Valjean', 0.47368421052631576),
 ('Marguerite', 0.02631578947368421),
 ('MmeDeR', 0.013157894736842105),
 ('Isabeau', 0.013157894736842105),
 ('Gervais', 0.013157894736842105),
 ('Tholomyes', 0.11842105263157894),
 ('Listolier', 0.09210526315789473),
 ('Fameuil', 0.09210526315789473),
 ('Blacheville', 0.09210526315789473),
 ('Favourite', 0.09210526315789473),
 ('Dahlia', 0.09210526315789473),
 ('Zephine', 0.09210526315789473),
 ('Fantine', 0.19736842105263158),
 ('MmeThenardier', 0.14473684210526316),
 ('Thenardier', 0.21052631578947367),
 ('Cosette', 0.14473684210526316),
 (

we can see that there is a tuple that has the highest degree of centrality which is ('Valjean', 0.47368421052631576).
so the character Valjean has the most influence in the network. We can verify that like so:

In [11]:
print("Max degree_centrality: ", max(list(nx.degree_centrality(G).values())))

Max degree_centrality:  0.47368421052631576


This makes sense since Valjean is the character than has the most connections. i.e. the only character that was almost present in every chapter of les miserables. We can verrify that by looking at the number of connections every node has. We'll see that the character Valjean has 36 connections, like so:

In [16]:
for n in G.nodes():
    print (n + '----->' + str(len(list(G.neighbors(n)))))

Myriel----->10
Napoleon----->1
MlleBaptistine----->3
MmeMagloire----->3
CountessDeLo----->1
Geborand----->1
Champtercier----->1
Cravatte----->1
Count----->1
OldMan----->1
Labarre----->1
Valjean----->36
Marguerite----->2
MmeDeR----->1
Isabeau----->1
Gervais----->1
Tholomyes----->9
Listolier----->7
Fameuil----->7
Blacheville----->7
Favourite----->7
Dahlia----->7
Zephine----->7
Fantine----->15
MmeThenardier----->11
Thenardier----->16
Cosette----->11
Javert----->17
Fauchelevent----->4
Bamatabois----->8
Perpetue----->2
Simplice----->4
Scaufflaire----->1
Woman1----->2
Judge----->6
Champmathieu----->6
Brevet----->6
Chenildieu----->6
Cochepaille----->6
Pontmercy----->3
Boulatruelle----->1
Eponine----->11
Anzelma----->3
Woman2----->3
MotherInnocent----->2
Gribier----->1
Jondrette----->1
MmeBurgon----->2
Gavroche----->22
Gillenormand----->7
Magnon----->2
MlleGillenormand----->7
MmePontmercy----->2
MlleVaubois----->1
LtGillenormand----->4
Marius----->19
BaronessT----->2
Mabeuf----->11
Enjolras---

### do the calculation manually

In [19]:
number_of_neighbors_Valjean = len(list(G.neighbors('Valjean')))
print("Total number of neighbors of Valjean: ", number_of_neighbors_Valjean)
print("Total number of nodes in the Graph is: ", len(G.nodes()) )
print("Degree Centrality of Valjeans is number_of_neighbors_Valjean/(len(G.nodes()) - 1)) = ", number_of_neighbors_Valjean/(len(G.nodes()) - 1))

Total number of neighbors of Valjean:  36
Total number of nodes in the Graph is:  77
Degree Centrality of Valjeans is number_of_neighbors_Valjean/(len(G.nodes()) - 1)) =  0.47368421052631576


**the manual calculation and the API yield the same result, that is 0.47368421052631576.**

## Let's do the graph

In [20]:
#visualizing the network with pyvis:
g=net.Network(height="800px", width="100%", notebook=True)
nxg = nx.Graph(G)
g.from_nx(nxg)
g.show("basic.html")
 
# notice that Valjean character is the one with the most connections. Clicking on Valjean node will activate it and 
#will show all the connections edges in bold.

# #visualizing the network with nxViz
# c = nv.CircosPlot(G)
# c.draw()
# plt.show()

Link to the video can be found in youtube at [this link](https://youtu.be/qKpMFhubQ5g)
